skorch_tabnet_classifier.ipynb

```diff
@@ -0,0 +1,1247 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 13,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class CrossEntropySparsityLoss(torch.nn.CrossEntropyLoss):\n",
+    "    def __init__(self, lambda_sparse=1e-3):\n",
+    "        super().__init__()\n",
+    "        self.lambda_sparse = lambda_sparse\n",
+    "        \n",
+    "    def forward(self, y_pred, y_true):\n",
+    "        output, M_loss = y_pred\n",
+    "\n",
+    "        loss = super().forward(output, y_true)\n",
+    "        \n",
+    "        # Add the overall sparsity loss\n",
+    "        loss -= self.lambda_sparse * M_loss\n",
+    "        \n",
+    "        return loss"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In terms of parametrization we just copy the parameters over from the tabnet example notebook.\n",
+    "Parameters that actually belong to the PyTorch module are now set directly via skorch's\n",
+    "`module__<param>` notation.\n",
+    "\n",
+    "We also use skorch's `LRScheduler`, `EarlyStopping`, `GradientNormClipping` and `*Scoring`\n",
+    "callbacks to re-implement the training regiment the original notebook uses.\n",
+    "\n",
+    "Note that for `EarlyStopping` to work we need a validation loss during training which we\n",
+    "only have once we add validation data. The difference to the original notebook is that\n",
+    "we supply this data at initialization time instead of training time via the `train_split`\n",
+    "parameter."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "torch.manual_seed(0)\n",
+    "\n",
+    "skorch_clf = SkorchTabModel(\n",
+    "    criterion=CrossEntropySparsityLoss,\n",
+    "    \n",
+    "    # Configure the module the same as in the tabnet example.\n",
+    "    module__input_dim=X_train.shape[-1],\n",
+    "    module__output_dim=infer_output_dim(y_train)[0],\n",
+    "    module__cat_idxs=cat_idxs,\n",
+    "    module__cat_dims=cat_dims,\n",
+    "    module__cat_emb_dim=1,\n",
+    "    module__mask_type='entmax',\n",
+    "    module__virtual_batch_size=128,\n",
+    "    \n",
+    "    optimizer=torch.optim.Adam,\n",
+    "    optimizer__lr=2e-2,\n",
+    "    \n",
+    "    batch_size=1024,\n",
+    "    iterator_train__num_workers=0,\n",
+    "    iterator_train__drop_last=False,\n",
+    "    iterator_valid__num_workers=0,\n",
+    "    iterator_valid__drop_last=False,\n",
+    "    \n",
+    "    callbacks=[\n",
+    "        skorch.callbacks.LRScheduler(\n",
+    "            policy=torch.optim.lr_scheduler.StepLR,\n",
+    "            step_size=50,\n",
+    "            gamma=0.9,\n",
+    "        ),\n",
+    "        skorch.callbacks.EarlyStopping(patience=20),\n",
+    "        skorch.callbacks.GradientNormClipping(gradient_clip_value=1.),\n",
+    "        skorch.callbacks.EpochScoring('roc_auc'),\n",
```

Comment:
This callback does not work for me with an error `AttributeError: 'SkorchTabModel' object has no attribute 'classes_'`.
Which version of skorch are you using?

Reply:
It does work if I add `self.classes_ = np.array([0, 1])` to the function `initialize_module`.

Reply:
@eduardocarvp I'm not sure that we want to have the notebook in CI, at least not mandatory. It could be nice to see if something breaks but not being able to merge without skorch working does not seem like a good idea.

I guess having the notebook as an example is a good option (it should be clearly stated like this at the beginning of the notebook)

Reply:
Yes I agree, that's why I didn't ask it on the review. I just meant that if we ever want to be fully compatible in the future why not, but for now I agree with you that just a notebook example is a better option.

Reply:
> This callback does not work for me with an error AttributeError: 'SkorchTabModel' object has no attribute 'classes_'. Which version of skorch are you using?

Interesting, I strongly suspect that this is related to a difference in sklearn versions. I used sklearn 0.23.2, what version did you test this with? My skorch version at the time was 0.10.1dev (from master). I have since tested it on the latest release (0.10.0) and did not run into the issue.

In any case, I changed the base class of my implementation to the `NeuralNetClassifier` class which infers the `classes_` attribute automatically.

Reply:
@eduardocarvp Can you tell me which sklearn version you tested this with? :)